### ISSUE TITLE
narrowing union of generic with `Any` regression

### ISSUE THREAD

**Issue by beauxq** [microsoft/pyright#8975]
This code checked with no errors in 1.1.379, but now reports an error with 1.1.380

```python
def foo(x: dict[Any, Any] | Any) -> TypeGuard[dict[int, Any]]:
    return isinstance(x, dict) and all(isinstance(k, int) for k in x)
```

This union with `Any` was a technique suggested to me by a maintainer of typeshed.

1.1.379:
https://pyright-play.net/?pyrightVersion=1.1.379&strict=true&reportMissingModuleSource=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcANFACqICmA4gK4CGIAJgFDvNXBTBhgAUADwBcUZkgDGMANpFShYgF0oAHwVwAlFAC0APgrV6TZtPFTpqGGTmLFw1lEdQQVGHRApMAZ1ReYDFAkqITIzGC0A5igGABsY-iQfFD8AoP4AazJLLV4QKHTMT0ENViA

1.1.380:
https://pyright-play.net/?pyrightVersion=1.1.380&strict=true&reportMissingModuleSource=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcANFACqICmA4gK4CGIAJgFDvNXBTBhgAUADwBcUZkgDGMANpFShYgF0oAHwVwAlFAC0APgrV6TZtPFTpqGGTmLFw1lEdQQVGHRApMAZ1ReYDFAkqITIzGC0A5igGABsY-iQfFD8AoP4AazJLLV4QKHTMT0ENViA

Was this change intentional?
If so, how would you recommend annotating this code?

**Comment by erictraut** [comment]
Version 1.1.380 contains some significant rework of the `isinstance` type narrowing code, including a bunch of bug fixes. In this case, the correct narrowed type is `dict[Any, Any] | dict[Unknown, Unknown]`.

In standard type checking mode, there are no errors reported in your code. In strict mode, there is a `reportUnknownVariableType` diagnostic reported, which makes sense because of the `Unknown` produced by narrowing. You can suppress that if you'd like using a `cast` call or a `# pyright: ignore` comment.

```python
def foo1(x: object) -> TypeGuard[dict[int, Any]]:
    return isinstance(x, dict) and all(isinstance(k, int) for k in cast(dict[Any, Any], x))

def foo2(x: object) -> TypeGuard[dict[int, Any]]:
    return isinstance(x, dict) and all(isinstance(k, int) for k in x)  # pyright: ignore[reportUnknownVariableType]
```
